bdk_core/src/io.rs

```diff
@@ -0,0 +1,249 @@
+use core::array::TryFromSliceError;
+#[cfg(feature = "std")]
+use std::convert::TryFrom;
+#[cfg(feature = "std")]
+use std::{ fmt, error::Error };
+#[cfg(feature = "std")]
+use std::io::{self, Read, Seek, Write};
+
+use bitcoin::{consensus::{self, encode}, BlockHash, Txid, Transaction, TxOut};
+
+use crate::{collections::*, Vec, tx_graph::{TxGraph, TxNode}, sparse_chain::{SparseChain, TxHeight, UpdateFailure } };
+
+/// Persistence error of [`SparseChain`] and [`TxGraph`]
+#[derive(Debug)]
+pub enum IOError {
+    /// wrapper for std::io::Error
+    Io(io::Error),
+    /// Value used for kind_byte is unknown
+    UnknownKind(u8),
+    /// consensus deserialize/serialize error
+    Consensus(consensus::encode::Error),
+    /// slice to array conversion error
+    SliceConversion(TryFromSliceError),
+    /// [`TxGraph`] update error
+    TxUpdateError(UpdateFailure)
+}
+
+#[cfg(feature = "std")]
+impl Error for IOError {}
+
+impl fmt::Display for IOError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl From<io::Error> for IOError {
+    fn from(error: std::io::Error) -> IOError {
+        IOError::Io(error)
+    }
+}
+
+impl From<encode::Error> for IOError {
+    fn from(error: encode::Error) -> IOError {
+        IOError::Consensus(error)
+    }
+}
+
+impl From<TryFromSliceError> for IOError {
+    fn from(error: TryFromSliceError) -> IOError {
+        IOError::SliceConversion(error)
+    }
+}
+
+impl From<UpdateFailure> for IOError {
+    fn from(error: UpdateFailure) -> IOError {
+        IOError::TxUpdateError(error)
+    }
+}
+
+const UNCONFIRMED_TX_HEIGHT: u32 = u32::MAX - 1;
+
+/// write SparseChain to disk for persistence
```

Comment:
Note this function doesn't necessarily write it to disk. Just anything that is `Write`.